src/client/Microsoft.Identity.Client/Http/Headers/WwwAuthenticateParameters.cs

```diff
@@ -0,0 +1,464 @@
+        /// <param name="cancellationToken">The cancellation token to cancel operation.</param>
+        /// <returns>WWW-Authenticate Parameters extracted from response to the unauthenticated call.</returns>
+        public static async Task<IReadOnlyList<WwwAuthenticateParameters>> CreateFromAuthenticationResponseAsync(HttpClient httpClient, string resourceUri, CancellationToken cancellationToken = default)
+        {
+            if (httpClient is null)
+            {
+                throw new ArgumentNullException(nameof(httpClient));
+            }
+            if (string.IsNullOrWhiteSpace(resourceUri))
+            {
+                throw new ArgumentNullException(nameof(resourceUri));
+            }
+
+            // call this endpoint and see what the header says and return that
+            HttpResponseMessage httpResponseMessage = await httpClient.GetAsync(resourceUri, cancellationToken).ConfigureAwait(false);
+            var wwwAuthParams = CreateFromAuthenticateHeaders(httpResponseMessage.Headers);
+            return wwwAuthParams;
+        }
+
+        /// <summary>
+        /// Create WWW-Authenticate parameters from the HttpResponseHeaders for each auth scheme.
+        /// </summary>
+        /// <param name="httpResponseHeaders">HttpResponseHeaders.</param>
+        /// <returns>The parameters requested by the web API.</returns>
+        /// <remarks>Currently it only supports the Bearer scheme</remarks>
+        public static IReadOnlyList<WwwAuthenticateParameters> CreateFromAuthenticateHeaders(
+            HttpResponseHeaders httpResponseHeaders)
+        {
+            List<WwwAuthenticateParameters> parameterList = new List<WwwAuthenticateParameters>();
+
+            foreach (AuthenticationHeaderValue wwwAuthenticateHeaderValue in httpResponseHeaders.WwwAuthenticate)
+            {
+                var parameters = CreateFromWwwAuthenticateHeaderValue(wwwAuthenticateHeaderValue.Parameter);
+                parameters.AuthScheme = wwwAuthenticateHeaderValue.Scheme;
+                parameterList.Add(parameters);
+            }
+
+            return parameterList;
+        }
+        #endregion Multi Scheme Api
+
+        /// <summary>
+        /// Creates parameters from the WWW-Authenticate string.
+        /// </summary>
+        /// <param name="wwwAuthenticateValue">String contained in a WWW-Authenticate header.</param>
+        /// <returns>The parameters requested by the web API.</returns>
+        public static WwwAuthenticateParameters CreateFromWwwAuthenticateHeaderValue(string wwwAuthenticateValue)
+        {
+            if (string.IsNullOrWhiteSpace(wwwAuthenticateValue))
+            {
+                throw new ArgumentNullException(nameof(wwwAuthenticateValue));
+            }
+            IDictionary<string, string> parameters;
+
+            var authValuesSplit = wwwAuthenticateValue.Split(new char[] { ' ' }, 2);
+
+            if (s_knownAuthenticationSchemes.Contains(authValuesSplit[0]))
```

Comment:
I was under the impression that we are parsing all schemes? Not sure where the spec landed on this.

Reply:
`The apis that return dictionaries should try to return a collection of all of the WWW-AuthenticateHeaders found, not just the ones we know about.`

Reply:
This is actually one of the obsolete apis. Also, this check for s_knownAuthenticationSchemes was used to check if the developer passed in the wwwAuthenticateValue with the auth scheme added to it which caused issues with parsing. The new apis parse the entire collection of WWW-Authentication headers.